### PR TITLE
drivers/adxl345: Initial support + saul

### PR DESCRIPTION
**Pull request by dylad**
Initial support for [ADXL345](http://www.analog.com/media/en/technical-documentation/data-sheets/ADXL345.pdf) accelerometer (I2C support only) + SAUL

### CODE REVIEW

File: drivers/include/adxl345.h [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+ * @param[out] data         the current acceleration data [in mg]
+ */
+void adxl345_read(adxl345_t *dev, adxl345_data_t *data);
+
+/**
+ * @brief   set ADXL345's interrupts configuration
+ *
+ * @param[in]  dev          device descriptor of accelerometer
+ */
+void adxl345_interrupt(adxl345_t *dev);
```
Comment (aabadie) [inline drivers/include/adxl345.h]: the function here and the ones above looks like internal implementation. I think they should be kept in the .c file.

Comment (aabadie) [inline drivers/include/adxl345.h]: except if you want to expose them. In this case, I would prefer a name like adxl1345_set_xxx.

File: tests/driver_adxl345/Makefile [hunk 2]
```diff
@@ -0,0 +1,9 @@
+APPLICATION = driver_adxl345
+include ../Makefile.tests_common
+
+FEATURES_REQUIRED = periph_i2c
```
Comment (aabadie) [inline tests/driver_adxl345/Makefile]: this is not required as adxl135 already imports it

File: tests/driver_adxl345/main.c [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+#include <stdio.h>
+
+#include "adxl345.h"
+#include "adxl345_params.h"
+#include "xtimer.h"
+
+#define SLEEP_DELAY 100 * 1000U
+
+static adxl345_t dev;
```
Comment (aabadie) [inline tests/driver_adxl345/main.c]: don't need to declare it static, it can go as a local variable in main.

File: tests/driver_adxl345/main.c [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+int main(void)
+{
+    adxl345_data_t data;
+
+    puts("ADXL345 test application");
+    printf("Initializing ADXL345 accelerometer at I2C_DEV(%i)... ",
+           adxl345_params->i2c);
+
+    if (adxl345_init(&dev, adxl345_params) == ADXL345_OK) {
```
Comment (aabadie) [inline tests/driver_adxl345/main.c]: I would prefer testing the error return code first (NOI2C,  NODEV) and display the corresponding reason.

**Comment by dylad** [comment]
Thanks for the comments @aabadie 
I'll commit the required changes soon and fix doxygen problems from Jenkins

File: tests/driver_adxl345/main.c (hunk 4 above)
Comment (dylad) [inline tests/driver_adxl345/main.c]: Have you any reference for this ? tests/drivers_lsm303dlhc do the same init

**Comment by dylad** [comment]
@aabadie I commit the requested changes.
Could you please have a look at it ?

File: drivers/adxl345/adxl345.c [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+#define ADDR                (dev->params.addr)
+
+int adxl345_init(adxl345_t *dev, const adxl345_params_t* params)
+{
+    uint8_t reg;
+
+    assert(dev && params);
+
+    /* write device descriptor */
+    memcpy(dev, params, sizeof(adxl345_params_t));
```
Comment (aabadie) [inline drivers/adxl345/adxl345.c]: you can use 
```c
dev->params = *params;
```

File: sys/auto_init/saul/auto_init_adxl345.c [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+/** @} */
+
+void auto_init_adxl345(void)
+{
+    for (unsigned i = 0; i < ADXL345_NUM; i++) {
+        LOG_DEBUG("[auto_init_saul] initializing adxl345 #%u\n", i);
+
+        if (adxl345_init(&adxl345_devs[i], &adxl345_params[i]) != ADXL345_OK) {
+            LOG_ERROR("[auto_init_saul] error initializing adxl345 #%u\n", i);
+            return;
```
Comment (aabadie) [inline sys/auto_init/saul/auto_init_adxl345.c]: I think it should be `continue` (this problem is also present in other places).

File: tests/driver_adxl345/main.c [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    }
+    else {
+        puts("[Failed]");
+        return -1;
+    }
+
+   while(1) {
+        adxl345_read(&dev, &data);
+        printf("Acceleration [in mg]: X axis:%d Y axis:%d Z axis:%d\n",
+                data.x,data.y,data.z);
```
Comment (aabadie) [inline tests/driver_adxl345/main.c]: use space after commas

File: drivers/adxl345/adxl345_saul.c [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+ * @}
+ */
+
+#include "saul.h"
+#include "adxl345.h"
+
+static int read_acc(void *dev, phydat_t *res)
+{
+    adxl345_t *d = (adxl345_t *)dev;
+    adxl345_read(d, (adxl345_data_t *)res);
```
Comment (aabadie) [inline drivers/adxl345/adxl345_saul.c]: use `res->val` to make it less confusing.

**Comment by dylad** [comment]
Update this PR with the latest review.

**Comment by dylad** [comment]
Many thanks @aabadie !

**Comment by aabadie** [comment]
All green but there's one remaining thing : can you squash in 2 commits:
* "drivers/adxl345: initial implementation + saul adaptation" : containing the code you added in drivers and sys
* "tests/adxl345: add functionnal test": containing your changes in the tests

Thanks !

**Comment by dylad** [comment]
@aabadie I've squashed into two commits as you requested.
Waiting for the green lights !

**Comment by aabadie** [comment]
I don't have the hardware for testing. Can you provide an output or something that proves it works (this is also important ;)) ?

**Comment by dylad** [comment]
@aabadie I'll share the results later.
Or if you willing to, I can lend you some ADXL345 module if you're on Paris. 
What should be better for you ?

**Comment by aabadie** [comment]
> Or if you're willing to, I can lend you some ADXL345 module if you're on Paris.
What should be better for you ?

Looks interesting :) Maybe we could meet in Bourg la Reine station (I also travel daily in RERB to Polytechnique). Send me a mail directly.

**Comment by aabadie** [comment]
Tested works, thanks @dylad.

All CIs are green, so go !

**Comment by haukepetersen** [comment]
Hej, nice work and thanks for the driver! Could you be however so nice to do some cleanup/enhancements: the concept of the device parameter struct is misinterpreted and is wasting memory in the current state. It would be nice if you could only leave the actual static configuration parameters in the `adxl345_params_t` struct and move all dynamic fields (e.g. `interrupt`, `full_res`, `scale_factor`) into the device descriptor `adxl345_t`?! 

Also there are still some minor things in the driver, e.g.
-  `#define ADXL345_PARAM_SCALE_FACTOR (3.9)` -> float to uint8_t assignment? Can be dropped anyway (see above)
- indention seems off at some points in the header files
- I guess the interface could be improved and ROM could be saved by merging the 4 set state functions (but this might be a subjective one)

**Comment by dylad** [comment]
@haukepetersen Thanks for your comments !
I'll submit another PR in a few days to include your remarks.
Don't hesitate to add more reviews for this driver.

**Comment by haukepetersen** [comment]
nice, thanks!

**Pull request merged**